### PR TITLE
[Agent] Support for the Agent to modify its own container resource limits

### PR DESCRIPTION
**Pull request by TomatoMr**
### This PR is for:

- Agent

### Support for the Agent to modify its own container resource limits
- Added max_millicpus configuration, where the effective limit is the smaller value between max_millicpus and max_cpus.
- Container resource limitations are no longer dependent on YAML settings and are uniformly issued by the server.
- The minimum granularity for CPU resource limitations is milli-cores.
- Deprecated the environment variable K8S_MEM_LIMIT_FOR_DEEPFLOW.
- Refactored environment.rs to differentiate handling for Linux and Windows.
#### Checklist
- [ ] Added unit test.
#### Backport to branches
- main

### CODE REVIEW

File: agent/Cargo.toml [hunk 1]
```diff
@@ -83,6 +83,7 @@ tonic = "0.8.1"
 envmnt = "0.10.4"
 wasmtime = "12.0.1"
 wasmtime-wasi = "12.0.1"
+bollard = "0.16.1"
```
Comment (rvql) [inline agent/Cargo.toml:86]: 排序

File: agent/src/config/handler.rs [hunk 2]
```diff
@@ -195,6 +198,8 @@ impl fmt::Debug for CollectorConfig {
 pub struct EnvironmentConfig {
     pub max_memory: u64,
     pub max_cpus: u32,
+    pub max_millicpus: u32,
+    pub limited_millicpus: u32, // Compatible with max_cpus and max_millicpus, take the smaller value in milli-cores.
```
Comment (rvql) [inline agent/src/config/handler.rs:202]: 这个字段看上去可以用一个函数实现，即
```
impl EnvironmentConfig {
    pub fn limited_millicpus(&self) -> u32
}
```

File: agent/Cargo.toml (hunk 1 above)
Comment (TomatoMr) [reply agent/Cargo.toml]: 修改了

File: agent/src/config/handler.rs (hunk 2 above)
Comment (TomatoMr) [reply agent/src/config/handler.rs]: 是的，修改了

Comment (rvql) [reply agent/src/config/handler.rs]: `max_cpu`还需要吗？

File: agent/src/config/config.rs [hunk 3]
```diff
@@ -1154,6 +1154,7 @@ pub struct RuntimeConfig {
     #[serde(skip)]
     pub enabled: bool,
     pub max_cpus: u32,
+    pub max_millicpus: u32,
```
Comment (rvql) [inline agent/src/config/config.rs:1157]: `max_cpu`还需要吗？

File: agent/src/utils/environment/linux.rs [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+
+const CORE_FILE_CONFIG: &str = "/proc/sys/kernel/core_pattern";
+const CORE_FILE_LIMIT: usize = 3;
+
+pub fn free_memory_check(_required: u64, _exception_handler: &ExceptionHandler) -> Result<()> {
+    return Ok(()); // fixme: The way to obtain free memory is different in earlier versions of Linux, which requires adaptation
+}
+
+pub fn kernel_check() {
+    {
```
Comment (rvql) [inline agent/src/utils/environment/linux.rs:51]: 为什么要多写一层大括号？

File: agent/src/utils/environment/linux.rs [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+    None
+}
+
+pub fn get_container_resource_limits() -> (u32, u64) {
+    if !running_in_container() {
+        return (0, 0);
+    }
+    let cpu_cgroups_files = [
+        "/sys/fs/cgroup/cpu.max", // If the container image uses cgroups v2, the format of the file is: {cfs_quota_us} {cfs_period_us}, for example: 100000 100000
+        "cat /sys/fs/cgroup/cpu/cpu.cfs_quota_us", // If the container image uses cgroups v1, the format of the file is: {cfs_quota_us}, for example: 100000
```
Comment (rvql) [inline agent/src/utils/environment/linux.rs:288]: cat?

File: agent/src/utils/guard.rs [hunk 6]
```diff
@@ -264,7 +264,7 @@ impl Guard {
                 return false;
             }
         };
-        (cpu_limit * 100) as f32 > cpu_usage
+        (cpu_limit / 10) as f32 > cpu_usage
```
Comment (rvql) [inline agent/src/utils/guard.rs:267]: 这里需要解释为什么除 10

File: agent/src/config/config.rs (hunk 3 above)
Comment (TomatoMr) [reply agent/src/config/config.rs:1156]: 确实可以去掉了，我改成了 max_millicpus

File: agent/src/utils/kubernetes_quantity.rs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+const SUFFIX_KI: &'static str = "Ki";
+const SUFFIX_MI: &'static str = "Mi";
+const SUFFIX_GI: &'static str = "Gi";
+const SUFFIX_TI: &'static str = "Ti";
+const SUFFIX_PI: &'static str = "Pi";
+const SUFFIX_EI: &'static str = "Ei";
+
+const INVALID_QUANTITY: &'static str = "invalid quantity";
+
+fn parse_quantity(q: &Quantity) -> Result<(f64, String), &'static str> {
```
Comment (rvql) [inline agent/src/utils/kubernetes_quantity.rs:49]: 写复杂了
1. 将字符串分成数值部分和单位部分。直接从字符串后往前看至多两个字符就能分开
2. 数值部分直接用 `f64` 就可以解析出来
3. 单位不用搞这么多常量，像这样就很具可读性了
```
match suffix {
    "M" => 10u64.pow(6),
    "Mi" => 1 << 20,
}
```
4. 函数的返回值应该是 `Result<(f64, &str), &'static str>`

File: agent/src/utils/kubernetes_quantity.rs [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        num.clone()
+    } else {
+        format!("{}.{}", num, denom)
+    };
+
+    let result = value.parse::<f64>().map_err(|_| INVALID_QUANTITY)?;
+    Ok((result * multiple as f64, suffix))
+}
+
+pub fn convert_to_mill_cpus(q: &Quantity) -> Result<u32, &'static str> {
```
Comment (rvql) [inline agent/src/utils/kubernetes_quantity.rs:111]: 直接把 `parse_quantity` 输出改成“毫”为单位就可以不用特殊处理

File: agent/src/config/config.rs [hunk 9]
```diff
@@ -1481,7 +1481,10 @@ impl TryFrom<trident::Config> for RuntimeConfig {
         let rc = Self {
             vtap_group_id: Default::default(),
             enabled: conf.enabled(),
-            max_cpus: conf.max_cpus(),
+            max_millicpus: conf
+                .max_millicpus
+                .unwrap_or(conf.max_cpus() * 1000)
+                .min(conf.max_cpus() * 1000), // Compatible with max_cpus and max_millicpus, take the smaller value in milli-cores.
```
Comment (rvql) [inline agent/src/config/config.rs:1487]: 应该没解决下发为 0 的情况
```suggestion
            max_millicpus: {
                let opt1 = conf.max_cpus() * 1000;
                let opt2 = conf.max_millicpus();
                if opt1 != 0 && opt2 != 0 {
                     opt1.min(opt2);
                } else {
                    opt1 | opt2
                }
            }
```

File: agent/src/config/config.rs [hunk 10]
```diff
@@ -1481,7 +1481,15 @@ impl TryFrom<trident::Config> for RuntimeConfig {
...
-            max_cpus: conf.max_cpus(),
+            max_millicpus: {
+                let max_cpus = conf.max_cpus() * 1000;
+                let max_millicpus = conf.max_millicpus.unwrap_or(max_cpus); // conf.max_millicpus may be None, handle the case where conf.max_millicpus is None
+                if max_cpus != 0 && max_millicpus != 0 {
+                    max_cpus.min(max_millicpus)
+                } else {
+                    max_cpus | max_millicpus
+                }
+            }, // Compatible with max_cpus and max_millicpus, take the smaller value in milli-cores.
```
Comment (rvql) [inline agent/src/config/config.rs:1492]: 注释位置调整一下

Comment (TomatoMr) [reply agent/src/config/config.rs]: done

**Review by sharang — APPROVED** [review]

**Review by rvql — APPROVED** [review]

**Pull request merged**